app/models/manager_refresh/graph.rb

```diff
@@ -40,9 +40,22 @@ def to_graphviz(layers: nil)
     def construct_graph!(nodes)
       self.nodes = nodes
       self.edges, self.fixed_edges = build_edges(nodes)
+      assert_edges_in_nodes!
+      assert_fixed_edges_in_edges!
       self
     end
 
+    def assert_edges_in_nodes!
+      edge_endpoints = edges.collect_concat { |e| e }
+      extra = edge_endpoints - nodes
+      raise "Graph has edge endpoints that are not among its nodes: #{extra}" unless extra.empty?
+    end
+
+    def assert_fixed_edges_in_edges!
+      extra = fixed_edges - edges
+      raise "Graph has fixed_edges that are not in edges: #{extra}" unless extra.empty?
```

Comment:
@Ladas is fixed_edges ⊂ edges a true invariant?  I think it's true on construction, but is it throughout a graph's life?  or is it misleading and better drop it?

Reply:
yes, edges should always be the superset https://github.com/Ladas/manageiq/blob/266cde6b9a68459c05c934192f2f2e713b33496d/app/models/manager_refresh/inventory_collection/graph.rb#L131

we probably don't need to assert it though, assert_edges_in_nodes should be enough